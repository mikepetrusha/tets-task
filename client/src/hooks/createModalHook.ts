import { FC, memo, useCallback, useRef } from 'react';

import { useDialogService } from '@/services/dialog';

export const createModalHook = <T>(Component: (props: T) => FC) => {
  return function useDialog() {
    const idRef = useRef<string>();

    const closeDialog = useCallback(() => {
      if (idRef.current) {
        useDialogService.getState().closeDialog(idRef.current);
      }
    }, []);

    const openDialog = useCallback(
      (props?: Omit<T, 'onClose'>) => {
        idRef.current = useDialogService
          .getState()
          .openDialog(memo(Component({ ...props, onClose: closeDialog } as T)));
      },
      [closeDialog]
    );

    return [openDialog, closeDialog] as const;
  };
};
