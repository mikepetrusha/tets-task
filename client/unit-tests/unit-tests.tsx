/* eslint-disable import/export */
import { render, RenderOptions } from '@testing-library/react';
import React, { ReactElement } from 'react';
import { RouterProvider } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';

import { router } from '@/router';

import { GlobalStyle } from '../src/config/theme';
import theme from '../src/config/theme/theme';

const AllProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyle />
      {children}
    </ThemeProvider>
  );
};

export const TestBrowserProvider = () => {
  return <RouterProvider router={router} />;
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: AllProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
