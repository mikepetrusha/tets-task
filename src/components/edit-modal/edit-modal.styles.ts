import { css, styled } from 'styled-components';

export const Overlay = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  z-index: 100;
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: rgba(0, 0, 0, 0.5);
`;

export const FormWrapper = styled.div`
  ${({ theme }) => {
    return css`
      padding: 20px;
      display: flex;
      flex-direction: column;
      gap: ${theme.gap?.medium};
    `;
  }}
`;

export const NameFormWrapper = styled.div`
  ${({ theme }) => {
    return css`
      display: flex;
      gap: ${theme.gap?.medium};

      @media (max-width: ${theme.breakpoints?.md}) {
        flex-direction: column;
      }
    `;
  }}
`;

export const ModalActions = styled.div`
  ${({ theme }) => {
    return css`
      padding: 8px 20px;
      display: flex;
      justify-content: space-between;
      background: ${theme.palette?.grey?.lightGrey};
    `;
  }}
`;

export const ButtonWrapper = styled.div`
  ${({ theme }) => {
    return css`
      display: flex;
      gap: ${theme.gap?.small};
    `;
  }}
`;
