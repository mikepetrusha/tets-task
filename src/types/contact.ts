export interface IContact {
  id: string;
  email: string;
  firstname: string;
  lastname: string;
}
