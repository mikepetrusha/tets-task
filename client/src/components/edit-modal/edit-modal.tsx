import { useEffect, useState } from 'react';
import { SubmitHandler, useForm } from 'react-hook-form';

import { createModalHook } from '@/hooks/createModalHook';
import { useContactsService } from '@/services/contacts';

import { Button, Input, Spinner } from '../ui';
import { Modal } from '../ui/modal';
import * as Styled from './edit-modal.styles';
import { IEditModalInput, IEditModalProps } from './edit-modal.types';

const EditModal = ({ onClose, title, contact }: IEditModalProps) => {
  const [isLoading, setIsLoading] = useState(true);
  const { editContact, deleteContact } = useContactsService();

  const handleEdit = (data: IEditModalInput) => {
    editContact(data);
    onClose();
  };

  const {
    register,
    handleSubmit,
    formState: { isDirty }
  } = useForm<IEditModalInput>();
  const onSubmit: SubmitHandler<IEditModalInput> = (data) =>
    handleEdit({ ...data, id: contact.id });

  useEffect(() => {
    setIsLoading(false);
  }, []);

  return (
    <>
      {isLoading ? (
        <Styled.Overlay>
          <Spinner />
        </Styled.Overlay>
      ) : (
        <Modal title={title} onClose={onClose} open>
          <form onSubmit={handleSubmit(onSubmit)}>
            <Styled.FormWrapper>
              <Styled.NameFormWrapper>
                <Input
                  defaultValue={contact.firstname}
                  {...register('firstname', { required: true })}
                  placeholder="Vorname*"
                />
                <Input
                  defaultValue={contact.lastname}
                  {...register('lastname', { required: true })}
                  placeholder="Nachname*"
                />
              </Styled.NameFormWrapper>
              <Input
                type="email"
                defaultValue={contact.email}
                {...register('email', { required: true })}
                placeholder="E-Mail*"
              />
            </Styled.FormWrapper>
            <Styled.ModalActions>
              <Button onClick={() => deleteContact(contact.id)} variant="destructive">
                LÃ–SCHEN
              </Button>
              <Styled.ButtonWrapper>
                <Button variant="outlined" onClick={onClose}>
                  ABBRECHEN
                </Button>
                <Button disabled={!isDirty} type="submit">
                  SPEICHERN
                </Button>
              </Styled.ButtonWrapper>
            </Styled.ModalActions>
          </form>
        </Modal>
      )}
    </>
  );
};

export const useEditModal = createModalHook<IEditModalProps>((props) => () => (
  <EditModal {...props} />
));
