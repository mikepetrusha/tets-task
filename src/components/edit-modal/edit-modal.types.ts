import { IContact } from '@/types';

export interface IEditModalProps {
  title: string;
  contact: IContact;
  onClose: () => void;
}

export interface IEditModalInput {
  id: string;
  email: string;
  firstname: string;
  lastname: string;
}
