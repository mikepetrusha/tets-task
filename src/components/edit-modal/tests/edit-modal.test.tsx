import renderer from 'react-test-renderer';

import { createModalHook } from '@/hooks/createModalHook';

import { fireEvent, render, screen } from '../../../../unit-tests';
import { EditModal } from '../edit-modal';
import { IEditModalProps } from '../edit-modal.types';

const testContact = {
  id: '1',
  firstname: 'test1',
  lastname: 'test1',
  email: 'test1@gmail.com'
};

it('renders correctly', () => {
  const tree = renderer
    .create(<EditModal onClose={() => console.log('test')} title="test" contact={testContact} />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

const useEditModalTest = createModalHook<IEditModalProps>((props) => () => (
  <EditModal {...props} />
));

const TestComp = () => {
  const [openEditModal, closeEditModal] = useEditModalTest();
  openEditModal();
  closeEditModal();
  return <div>test</div>;
};
it('renders test correctly', () => {
  const tree = renderer.create(<TestComp />).toJSON();

  expect(tree).toMatchSnapshot();
});

describe('CreateModal', () => {
  it('should render modal', () => {
    render(<EditModal onClose={() => console.log('test')} title="test" />);

    const test = screen.getByText('LÃ–SCHEN');
    fireEvent.click(test);
  });
});
