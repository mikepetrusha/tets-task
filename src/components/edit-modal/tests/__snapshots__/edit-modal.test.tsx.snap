// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-ktwOfi dSPMWs"
>
  <div
    className="sc-eTNRI eHkrbZ"
    data-testid="test-spinner"
  >
    <div
      className="sc-brSamD eQuigN"
    />
  </div>
</div>
`;

exports[`renders test correctly 1`] = `
<div>
  test
</div>
`;
